src/confluent_kafka/src/confluent_kafka.c

```diff
@@ -1461,6 +1461,59 @@ static void throttle_cb (rd_kafka_t *rk, const char *broker_name, int32_t broker
         CallState_resume(cs);
 }
 
+static int32_t partitioner_cb (const rd_kafka_topic_t *rkt,
+			       const void *keydata,
+			       size_t keylen,
+			       int32_t partition_cnt,
+			       void *rkt_opaque,
+			       void *msg_opaque) {
+
+	PyObject *callback = rkt_opaque;
+	PyObject *result = NULL, *args = NULL;
+	PyObject *keydataobj = NULL;
+
+	/* This is only used for testing */
+	if (callback != NULL) {
+		PyObject *ks;
+		PyObject *ks8 = NULL;
+		PyObject *vs = NULL, *vs8 = NULL;
+		const char *k = NULL;
+		printf("Jing Liu callback before cfl_PyObject_Unistr\n");
+		ks = cfl_PyObject_Unistr(callback);
```

Comment:
Since the code keep failing at line 1504 below while running the PyObject_CallObject, I added this part for printing and verifying the function returned back from librdkafka, found that it failed here ```ks = cfl_PyObject_Unistr(callback);```. But I did the same printing and verifying before set the callback to librakafka at line 2164 below, found everything works as expected.

Reply:
The crash is most likely because you're accessing the Python interpreter without holding the GIL. You need to acquire the GIL before calling any Py* function, and release it before returning from this function.